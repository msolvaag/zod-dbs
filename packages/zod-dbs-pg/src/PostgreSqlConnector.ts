import { DatabaseConnector, logDebug, sql } from 'zod-dbs-core';

import type {
  ZodDbsConnectionConfig,
  ZodDbsConnectorConfig,
  ZodDbsRawColumnInfo,
} from 'zod-dbs-core';

import { createClient } from './client.js';

/**
 * Default connector to interact with PostgreSQL database and retrieve schema information.
 * Supports postgresql version 9.3 and above.
 */
export class PostgreSqlConnector extends DatabaseConnector {
  async createClient(options: ZodDbsConnectionConfig) {
    return await createClient(options);
  }

  public override async fetchSchemaInfo(
    config: ZodDbsConnectorConfig
  ): Promise<ZodDbsRawColumnInfo[]> {
    const { schemaName = 'public' } = config;

    config.onProgress?.('connecting');
    const client = await this.createClient(config);
    await client.connect();

    config.onProgress?.('fetchingSchema');
    logDebug(`Retrieving schema information for schema '${schemaName}'`);

    try {
      const res = await client.query<ZodDbsRawColumnInfo[]>(
        sql`
          SELECT
            c.relname AS "tableName",
            a.attname AS "name",
            pg_get_expr(d.adbin, d.adrelid) AS "defaultValue",
            t.typname AS "dataType",
            NOT a.attnotnull AS "isNullable",
            CASE 
              WHEN a.atttypmod > 0 THEN a.atttypmod - 4
              ELSE NULL
            END AS "maxLen",
            t.typname AS "dataType",
            checks."checkConstraints",
            col_description(c.oid, a.attnum) AS "description",
            CASE 
              WHEN c.relkind = 'r' THEN 'table'
              WHEN c.relkind = 'v' THEN 'view'
              WHEN c.relkind = 'm' THEN 'materialized_view'
              WHEN c.relkind = 'f' THEN 'foreign_table'
              ELSE 'unknown'
            END AS "tableType"
          FROM pg_class c
          JOIN pg_namespace n ON n.oid = c.relnamespace
          JOIN pg_attribute a ON a.attrelid = c.oid
          JOIN pg_type t ON t.oid = a.atttypid
          LEFT JOIN pg_attrdef d ON d.adrelid = c.oid AND d.adnum = a.attnum
          LEFT JOIN LATERAL (
            SELECT json_agg(json_build_object('checkClause', pg_get_constraintdef(pgc.oid))) AS "checkConstraints"
            FROM pg_constraint pgc
            WHERE pgc.conrelid = c.oid
              AND pgc.contype = 'c'
              AND pgc.conkey @> ARRAY[a.attnum]
          ) AS checks ON TRUE
          WHERE n.nspname = $1
            AND c.relkind IN ('r', 'v', 'm', 'f')
            AND a.attnum > 0
            AND NOT a.attisdropped
          ORDER BY c.relname, a.attnum;
        `,
        [schemaName]
      );

      logDebug(`Retrieved ${res.length} columns from schema '${schemaName}'`);

      return res.map((row) => ({ ...row, schemaName }));
    } finally {
      await client.end();
    }
  }
}
