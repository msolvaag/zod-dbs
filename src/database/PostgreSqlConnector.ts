import { Client } from 'pg';

import type {
  ZodPgConfig,
  ZodPgConnectionConfig,
  ZodPgRawColumnInfo,
} from '../types.js';

import { logDebug } from '../utils/debug.js';
import { sql } from '../utils/sql.js';
import { createConnectionString } from './client.js';
import { DatabaseConnector } from './DatabaseConnector.js';

/**
 * Default connector to interact with PostgreSQL database and retrieve schema information.
 * Supports postgresql version 9.3 and above.
 */
export class PostgreSqlConnector extends DatabaseConnector {
  protected override createClient = (options: ZodPgConnectionConfig) => {
    return new Client({
      connectionString: createConnectionString(options),
      ssl: options.ssl,
      application_name: 'zod-pg',
    });
  };

  protected override async fetchSchemaInfo(
    config: ZodPgConfig
  ): Promise<ZodPgRawColumnInfo[]> {
    const { schemaName = 'public' } = config;

    const client = this.createClient(config.connection);
    await client.connect();

    logDebug(`Retrieving schema information for schema '${schemaName}'`);

    try {
      const res = await client.query<ZodPgRawColumnInfo>(
        sql`
          SELECT
            c.relname AS "tableName",
            a.attname AS "name",
            pg_get_expr(d.adbin, d.adrelid) AS "defaultValue",
            t.typname AS "dataType",
            NOT a.attnotnull AS "isNullable",
            CASE 
              WHEN a.atttypmod > 0 THEN a.atttypmod - 4
              ELSE NULL
            END AS "maxLen",
            t.typname AS "dataType",
            checks."checkConstraints",
            col_description(c.oid, a.attnum) AS "description",
            CASE 
              WHEN c.relkind = 'r' THEN 'table'
              WHEN c.relkind = 'v' THEN 'view'
              WHEN c.relkind = 'm' THEN 'materialized_view'
              WHEN c.relkind = 'f' THEN 'foreign_table'
              ELSE 'unknown'
            END AS "tableType"
          FROM pg_class c
          JOIN pg_namespace n ON n.oid = c.relnamespace
          JOIN pg_attribute a ON a.attrelid = c.oid
          JOIN pg_type t ON t.oid = a.atttypid
          LEFT JOIN pg_attrdef d ON d.adrelid = c.oid AND d.adnum = a.attnum
          LEFT JOIN LATERAL (
            SELECT json_agg(json_build_object('checkClause', pg_get_constraintdef(pgc.oid))) AS "checkConstraints"
            FROM pg_constraint pgc
            WHERE pgc.conrelid = c.oid
              AND pgc.contype = 'c'
              AND pgc.conkey @> ARRAY[a.attnum]
          ) AS checks ON TRUE
          WHERE n.nspname = $1
            AND c.relkind IN ('r', 'v', 'm', 'f')
            AND a.attnum > 0
            AND NOT a.attisdropped
          ORDER BY c.relname, a.attnum;
        `,
        [schemaName]
      );

      logDebug(
        `Retrieved ${res.rows.length} columns from schema '${schemaName}'`
      );

      return res.rows.map((row) => ({
        ...row,
        schemaName,
      }));
    } finally {
      await client.end();
    }
  }
}
