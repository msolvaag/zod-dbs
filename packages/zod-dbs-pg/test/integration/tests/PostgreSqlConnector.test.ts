import { PostgreSqlConnector } from '../../../src/PostgreSqlConnector.js';
import {
  getClientConnectionString,
  setupTestDb,
  teardownTestDb,
  TestDbContext,
} from '../testDbUtils.js';

let ctx: TestDbContext;

beforeAll(async () => {
  ctx = await setupTestDb();
});

afterAll(async () => {
  await teardownTestDb(ctx);
});

it('returns schema information', async () => {
  const connector = new PostgreSqlConnector();

  const info = await connector.getSchemaInformation({
    connectionString: getClientConnectionString(),
  });

  expect(info).toBeDefined();
  expect(info.tables).toBeDefined();
  expect(info.tables).toHaveLength(19);

  const userTable = info.tables.find((t) => t.name === 'users')!;
  expect(userTable).toBeDefined();
  expect(userTable.columns).toBeDefined();
  expect(userTable.columns).toHaveLength(8);

  expect(userTable).toMatchSnapshot('userTable');
});
