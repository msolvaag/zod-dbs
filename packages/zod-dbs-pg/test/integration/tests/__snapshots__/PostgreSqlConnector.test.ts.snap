// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns schema information > userTable 1`] = `
{
  "columns": [
    {
      "checkConstraints": null,
      "dataType": "int4",
      "defaultValue": "nextval('users_id_seq'::regclass)",
      "description": null,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isOptional": false,
      "isSerial": true,
      "isWritable": false,
      "maxLen": undefined,
      "name": "id",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "int",
    },
    {
      "checkConstraints": null,
      "dataType": "varchar",
      "defaultValue": null,
      "description": null,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isOptional": false,
      "isSerial": false,
      "isWritable": true,
      "maxLen": 100,
      "name": "name",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "checkConstraints": null,
      "dataType": "varchar",
      "defaultValue": null,
      "description": null,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isOptional": true,
      "isSerial": false,
      "isWritable": true,
      "maxLen": 255,
      "name": "email",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "checkConstraints": null,
      "dataType": "varchar",
      "defaultValue": "'active'::character varying",
      "description": null,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isOptional": true,
      "isSerial": false,
      "isWritable": true,
      "maxLen": 20,
      "name": "status",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "checkConstraints": null,
      "dataType": "jsonb",
      "defaultValue": null,
      "description": null,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isOptional": true,
      "isSerial": false,
      "isWritable": true,
      "maxLen": undefined,
      "name": "profile",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "json",
    },
    {
      "checkConstraints": null,
      "dataType": "timestamptz",
      "defaultValue": "now()",
      "description": null,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isOptional": true,
      "isSerial": false,
      "isWritable": true,
      "maxLen": undefined,
      "name": "created_at",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "date",
    },
    {
      "checkConstraints": [
        {
          "checkClause": "CHECK ((roles <@ ARRAY['admin'::text, 'editor'::text, 'viewer'::text]))",
        },
      ],
      "dataType": "_text",
      "defaultValue": null,
      "description": null,
      "enumValues": [
        "admin",
        "editor",
        "viewer",
      ],
      "isArray": true,
      "isEnum": true,
      "isNullable": true,
      "isOptional": true,
      "isSerial": false,
      "isWritable": true,
      "maxLen": undefined,
      "name": "roles",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "checkConstraints": null,
      "dataType": "_timestamp",
      "defaultValue": null,
      "description": null,
      "isArray": true,
      "isEnum": false,
      "isNullable": true,
      "isOptional": true,
      "isSerial": false,
      "isWritable": true,
      "maxLen": undefined,
      "name": "dates",
      "schemaName": "public",
      "tableName": "users",
      "tableType": "table",
      "type": "date",
    },
  ],
  "name": "users",
  "schemaName": "public",
  "type": "table",
}
`;
